crates/ruff/src/rules/flake8_pyi/rules/duplicate_union_member.rs

```diff
@@ -1,89 +1,79 @@
-pub(crate) fn duplicate_union_member(checker: &mut Checker, expr: &Expr) {
-    let mut seen_nodes = FxHashSet::default();
-    traverse_union(&mut seen_nodes, checker, expr, None);
-}
+pub(crate) fn duplicate_union_member<'a>(checker: &mut Checker, expr: &'a Expr) {
+    let mut seen_nodes: std::collections::HashSet<ComparableExpr<'_>, _> = FxHashSet::default();
+    let mut diagnostics: Vec<Diagnostic> = Vec::new();
 
-fn traverse_union<'a>(
-    seen_nodes: &mut FxHashSet<ComparableExpr<'a>>,
-    checker: &mut Checker,
-    expr: &'a Expr,
-    parent: Option<&'a Expr>,
-) {
-    // The union data structure usually looks like this:
-    //  a | b | c -> (a | b) | c
-    //
-    // However, parenthesized expressions can coerce it into any structure:
-    //  a | (b | c)
-    //
-    // So we have to traverse both branches in order (left, then right), to report duplicates
-    // in the order they appear in the source code.
-    if let Expr::BinOp(ast::ExprBinOp {
-        op: Operator::BitOr,
-        left,
-        right,
-        range: _,
-    }) = expr
-    {
-        // Traverse left subtree, then the right subtree, propagating the previous node.
-        traverse_union(seen_nodes, checker, left, Some(expr));
-        traverse_union(seen_nodes, checker, right, Some(expr));
-    }
+    // Adds a member to `literal_exprs` if it is a `Literal` annotation
+    let mut check_for_duplicate_members = |expr: &'a Expr, parent: Option<&'a Expr>| {
+        // If we've already seen this union member, raise a violation.
+        if !seen_nodes.insert(expr.into()) {
+            let mut diagnostic = Diagnostic::new(
+                DuplicateUnionMember {
+                    duplicate_name: checker.generator().expr(expr),
+                },
+                expr.range(),
+            );
+            if checker.patch(diagnostic.kind.rule()) {
+                // Delete the "|" character as well as the duplicate value by reconstructing the
+                // parent without the duplicate.
 
-    // If we've already seen this union member, raise a violation.
-    if !seen_nodes.insert(expr.into()) {
-        let mut diagnostic = Diagnostic::new(
-            DuplicateUnionMember {
-                duplicate_name: checker.generator().expr(expr),
-            },
-            expr.range(),
-        );
-        if checker.patch(diagnostic.kind.rule()) {
-            // Delete the "|" character as well as the duplicate value by reconstructing the
-            // parent without the duplicate.
+                // SAFETY: impossible to have a duplicate without a `parent` node.
+                let parent = parent.expect("Parent node must exist");
 
-            // SAFETY: impossible to have a duplicate without a `parent` node.
-            let parent = parent.expect("Parent node must exist");
+                // If the parent node is not a `BinOp` we will not perform a fix
+                if let Expr::BinOp(ast::ExprBinOp { left, right, .. }) = parent {
```

Comment:
I think this could be written as `if let Some(Expr::BinOp(...)) = parent` to avoid the `.expect` above.

Reply:
This is a little awkward actually we still need to have an `.unwrap()` when we call `range()` unless we want to write two `if let` statements? Perhaps I'm missing something. It is kind of nice to panic here because the traversal code is _broken_ if this invariant does not hold but 🤷‍♀️ I can see both sides.